### PR TITLE
Fix location issue

### PR DESCRIPTION
**Pull request by CS6**
High - 使用者 enable 後的第一次紀錄，GPS 權限要要求視窗跳出後，尚未同意權限前，會跳出 Error 訊息 => 會重複詢問直到使用者做出決定 
High - 讓使用者可於設定中決定是否 enable GPS 收集，並且不 enable GPS 也可使用 (類似 Capture) =>: 預設不啟用啟用後開始詢問使用者權限
Medium - 設定頁面中 “Other info” 的 About 改為 Contace US (link 到 Numbers Contact US 頁面)

image

### CODE REVIEW

File: src/assets/i18n/en.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
   "splitter": ", ",
   "title": {
     "QRScanner": "Scan QR Code",
-    "about": "About",
+    "about": "Contace US",
```
Comment (shc261392) [inline src/assets/i18n/en.json]:
```suggestion
    "about": "Contact Us",
```

typo

File: src/app/tabs/settings/settings.page.ts [hunk 2]
```diff
@@ -144,8 +148,14 @@ export class SettingsPage implements OnInit, OnDestroy {
...
   }
 
+  locationSelected(event: CustomEvent): void {
+    this.updateFromPage.next({ enableLocation: event.detail.value });
+    this.proofService.setLocationInfoCollection(event.detail.value);
+  }
+
   fontSizeSelected(event: CustomEvent): void {
     this.updateFromPage.next({ fontSize: event.detail.value });
+    console.log('event.detail.value', event.detail.value);
```
Comment (shc261392) [inline src/app/tabs/settings/settings.page.ts]: Remove the log for debugging

File: src/app/tabs/settings/settings.page.html [hunk 3]
```diff
@@ -73,14 +73,25 @@ <h3>{{ 'title.selectRecordMode' | translate }}</h3>
...
-        <ion-select *ngIf="showSelects" [value]="userDataView.fontSize" (ionChange)="fontSizeSelected($event)"
+        <ion-select [value]="(userDataView.fontSize) ? userDataView.fontSize : fontSizes[0]" (ionChange)="fontSizeSelected($event)"
           [cancelText]="'title.cancel' | translate" [okText]="'title.confirm' | translate">
           <ion-select-option *ngFor="let fontsize of fontSizes" [value]="fontsize">
             {{ 'title.' + fontsize | translate }}
           </ion-select-option>
         </ion-select>
       </ion-item>
-
+      <ion-item lines="inset">
```
Comment (shc261392) [inline src/app/tabs/settings/settings.page.html:83]: `ion-toggle` may be a better choice for this option. I'd suggest use `ion-toggle` instead, but if there isn't enough time to implement it or if it's a UI component requested by PO then it's fine to stick to the current solution for now.

File: src/app/core/services/proof.service.ts [hunk 4]
```diff
@@ -13,11 +13,17 @@ import { GeolocationService } from './geolocation.service';
...
 })
 export class ProofService {
+  enableGeolocation: string = 'disable'; 
 
   constructor(
     private readonly geolocationService: GeolocationService,
   ) { }
 
+  async setLocationInfoCollection(enable: string) {
+    console.log("setLocationInfoCollection",enable);
```
Comment (shc261392) [inline src/app/core/services/proof.service.ts]: Remove the debugging log

File: src/app/shared/components/add-record/add-record.component.ts [hunk 5]
```diff
@@ -114,12 +114,22 @@ export class AddRecordComponent implements OnInit, OnDestroy {
...
         takeUntil(this.destroy$),
       );
+    } else {
+      return this.proofService.createDisableLocationProof()
+      .pipe(
+        tap(proof => this.updateRecordProof(proof)),
+        tap(() => this.proofStatus = ProofStatus.COMPLETE),
+        takeUntil(this.destroy$),
+      );
+    }
```
Comment (shc261392) [inline src/app/shared/components/add-record/add-record.component.ts]:
```suggestion
    } else {
      const proof = this.proofService.createProofWithoutLocation();
      this.updateRecordProof(proof);
      this.proofStatus = ProofStatus.COMPLETE);
    }
```

No need to use observable here

File: src/app/core/services/proof.service.ts [hunk 6]
```diff
@@ -28,6 +34,14 @@ export class ProofService {
       );
   }
 
+
+  createDisableLocationProof(): Observable<Proof> {
```
Comment (shc261392) [inline src/app/core/services/proof.service.ts]: The function isn't needed (see the comment in `add-record.component.ts`)

File: src/app/core/services/proof.service.ts [hunk 7]
```diff
@@ -13,11 +13,17 @@ import { GeolocationService } from './geolocation.service';
...
 export class ProofService {
+  enableGeolocation: string = 'disable'; 
 
   constructor(
     private readonly geolocationService: GeolocationService,
   ) { }
 
+  async setLocationInfoCollection(enable: string) {
+    console.log("setLocationInfoCollection",enable);
+    return this.enableGeolocation=enable;
```
Comment (shc261392) [inline src/app/core/services/proof.service.ts]:
```suggestion
    this.enableGeolocation = enable;
```
No need to return anything (the return value is not used)

File: src/app/core/services/proof.service.ts [hunk 8]
```diff
@@ -13,11 +13,17 @@ import { GeolocationService } from './geolocation.service';
   providedIn: 'root'
 })
 export class ProofService {
+  enableGeolocation: string = 'disable';
```
Comment (shc261392) [inline src/app/core/services/proof.service.ts]:
```suggestion
  enableGeolocation = 'disable'; 
```
The initialization already make the variable inferenced as string type, so there's no need for explicit type declaration.

File: src/app/core/services/geolocation.service.ts [hunk 9]
```diff
@@ -26,7 +26,17 @@ export class GeolocationService {
 
   getPosition(useCache = true): Observable<GeolocationPosition> {
     const cache = (this.isCachedPositionValid() && useCache);
-    const position$ = (cache) ? of(this.cachedPosition) : from(Geolocation.getCurrentPosition(this.defaultGeolocationOptions));
+
+    const getPosition = (): Promise<GeolocationPosition> =>
+      Geolocation
+        .getCurrentPosition(this.defaultGeolocationOptions)
+        .catch<GeolocationPosition>((error) => (error.message as string).includes('0')
```
Comment (shc261392) [inline src/app/core/services/geolocation.service.ts]: The meaning of `(error.message as string).includes('0')` is unclear. It's confusing to people who read the code but don't know what error you are catching, so it's better to rewrite the code in a way that it's easier to understand. Or if it's difficult to be expressive enough in code, write some comments to explain it.

And it doesn't seem necessary to create another function if all you want to do is catching errors. You can catch error using RxJS pipe operator `catchError` chained after `position$`, for example

```
return position$
  .pipe(
    catchError(error => of(undefined)),
```

The above example will  send `undefined` to be handled by the next operator in the pipe if position$ emits error.

**Review by shc261392 — CHANGES_REQUESTED** [review]

File: src/app/tabs/settings/settings.page.html (hunk 3 above)
Comment (CS6) [reply src/app/tabs/settings/settings.page.html:83]: This is the result of discussion with PO

**Review by shc261392 — APPROVED** [review]

**Pull request merged**